### DISCUSSION TITLE
[Q&A] TODOs

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#58] [CHAPTER 5 스트림 활용 - Q&A]
### b값 생성 부분
```java
IntStream.rangeClosed(1, 100)  // IntStream<>
                 .filter(b-> Math.sqrt(a*a + b*b) % 1 == 0 )
                 .boxed() // Stream<Integer> 
                 .map(b-> new int[]{a, b, (int) Math.sqrt(a*a + b*b)}); // Stream<int[]> ?
-------------------------------------------------------------------
IntStream.rangeClosed(1, 100)
                 .filter(b-> Math.sqrt(a*a + b*b) % 1 == 0 )
                 .mapToObj(b-> new int[]{a, b, (int) Math.sqrt(a*a + b*b)});
```

책에서는
> 1. Filter 연산 다음에 rangeClosed 가 반환한 IntStream 을 boxed을 이용해서 Stream<Integer>
로 복원했다. map은 스트림의 각 요소를 int 배열로 변환하기 때문이다. 
> 2. IntStream 의 map 메서드는 스트림의 각 요소로 int가 반환될 것을 기대하지만 이는 우리가 원하는 연산이 아니다. 
> 3. 개체값 스트림을 반환하는 IntStream의 mapToObj 메서드를 이용해서 이 코드를 재구현할 수 있다.

라고 한다.

### 2번의 말이 무슨 말인가?
1. IntStream을 boxed한 이유가, boxed없이 IntStream.map()을 할경우 map 안에서 int-> int 형태로 사용해야하는데, 
우리가 원하는 것은 int -> int[] 이기 때문에 Stream의 일반적인 map() 을 사용하기 위해 boxed().map() 을 사용했다 라는 말인가?
   - map 메서드는 스트림의 각 요소로 int가 반환해야 하는가?
   - IntStream.map() 과 Stream.map()가 정해진 리턴값이 있는가?

2. IntStream.boxed().map 메서드는 스트림의 각 요소로 int가 반환될 것을 기대하지만 우리가 원하는건 int[] 반환이기 때문에 mapToObj 를 사용했다는 말인가??

### boxed().map() 과 mapToObj()의 차이는 무엇일까?

**Comment by BvrPark** [comment]
### 2번의 말이 무슨 말인가?
**1. IntStream을 boxed한 이유가, boxed없이 IntStream.map()을 할경우 map 안에서 int-> int 형태로 사용해야하는데,
우리가 원하는 것은 int -> int[] 이기 때문에 Stream의 일반적인 map() 을 사용하기 위해 boxed().map() 을 사용했다 라는 말인가?**

-> 네 그 말 같습니다. IntStream.map()을 하면 int -> int 형태로 변환이 되어서 우리가 원하는 값인 int[]가 나오지 않기 때문에
책에서 그렇게 서술한 것 같습니다.
그래서 boxed() 메서드로 Stream<Integer>로 만든 다음에 map()을 사용해서 Stream<int[]>의 형태로 만들어야 되는 것 같습니다.

그리고 IntSteam.map()은 항상 int를 반환하는 것 같고 Stream.map()은 Stream에서 어떤 반환 값을 받냐에 따라서 다른 반환 값이 나오는 것 같습니다.

**2. IntStream.boxed().map 메서드는 스트림의 각 요소로 int가 반환될 것을 기대하지만 우리가 원하는건 int[] 반환이기 때문에 mapToObj 를 사용했다는 말인가??**
-> 아니요 IntStream.boxed().map()은 int[]을 반환하는 것이 맞고 mapToObj고 int[]를 반환하는 것이 맞습니다.

**3. boxed().map() 과 mapToObj()의 차이는 무엇일까?**
-> boxed().map()은 먼저 IntStream을 Stream<Integer>로 변환(boxed 메서드)한 다음, Stream<Integer>의 map() 메소드를 사용하는 반면, mapToObj()는 IntStream에 있는 메서드 이므로 int 값을 받아서 다양한 타입의 반환값이 나옵니다.
